src/main/java/network/brightspots/rcv/ResultsWriter.java

```diff
@@ -575,19 +569,41 @@ String writeGenericCvrCsv(
       // print header:
       // ContestId, TabulatorId, BatchId, RecordId, Precinct, Precinct Portion, rank 1 selection,
       // rank 2 selection, ... rank maxRanks selection
+      csvPrinter.print("Source Filepath");
       csvPrinter.print("Contest Id");
       csvPrinter.print("Tabulator Id");
       csvPrinter.print("Batch Id");
       csvPrinter.print("Record Id");
       csvPrinter.print("Precinct");
       csvPrinter.print("Precinct Portion");
-      for (int rank = 1; rank <= numRanks; rank++) {
+
+      // Get the max maxRankingsAllowed across all perSourceDataForCsv
+      int maxRankingAcrossSources = 0;
+      for (PerSourceDataForCsv sourceData : perSourceDataForCsv) {
+        maxRankingAcrossSources = Math.max(sourceData.maxRankingsAllowed, maxRankingAcrossSources);
+      }
+
+      for (int rank = 1; rank <= maxRankingAcrossSources; rank++) {
         String label = String.format("Rank %d", rank);
         csvPrinter.print(label);
       }
       csvPrinter.println();
+
+      // While the cast vote records are in a flattened list,
+      // we can use the PerSourceDataForCsv to determine which source each record came from.
+      int currentSourceIndex = 0;
+      PerSourceDataForCsv currentSourceData = perSourceDataForCsv.get(currentSourceIndex);
+
       // print rows:
-      for (CastVoteRecord castVoteRecord : castVoteRecords) {
+      for (int i = 0; i < castVoteRecords.size(); i++) {
+        if (i > currentSourceData.lastIndexInCvrList) {
+          // we've moved on to a new contest, so we need to switch to the next source
+          currentSourceIndex++;
```

Comment:
This feels a tiny bit fragile even if it's working right now because it assumes that the the ordering of the sources and the CVRs is stable.
 
Would it be more robust to store the source ID in the CVR object so we don't have to infer it from the CVR's index?

Reply:
Yes, though it would add an unnecessary sort -- how do you feel about this update: assume it's stable, but verify and throw a halting exception if it's not?

Reply:
I guess there's two things we need to verify:
1. `List<PerSourceDataForCsv> perSourceDataForCsv` hasn't had its order changed, since the field `lastIndexInCvrList` assumes its order is constant
2. `List<PerSourceDataForCsv>` and `List<CastVoteRecord>` are "in sync" with each other

My change addresses (1) but not (2).

I'm hesitant to add the source index to the CastVoteRecord because it would require being passed from the TabulatorSession down to the Reader and be implemented in each Reader individually. I can see it perhaps being useful to actually include a reference to the Source itself (but not its index?) in `CastVoteRecord`, which could even be used in lieu of `computedId` if we also included a line number. That feels broader than needed for the scope of this PR, though.

My justification is this: `perSourceDataForCsv` is explicitly and exclusively a structure intended to help index into a `List<CastVoteRecord>`. I've added a comment to specify this above writeRctabCvrCsv.

Reply:
OK, this all sounds fine for now.